Scan Accessory Error

Scan Accessory Error
<img width="1162" alt="图片" src="https://user-images.githubusercontent.com/72076317/149456256-423869cb-4ecc-4dc6-a373-c2c58675e94f.png">

error.log
```
2022-01-14T05:23:26Z [DEBUG] [/pkg/scan/job.go:374]: registration:
2022-01-14T05:23:26Z [INFO] [/pkg/scan/job.go:385]: {
  "uuid": "8efa5bcf-74e1-11ec-a02c-0242ac140004",
  "name": "Trivy",
  "description": "The Trivy scanner adapter",
  "url": "https://trivy-adapter:8443",
  "disabled": false,
  "is_default": true,
  "health": "healthy",
  "auth": "",
  "access_credential": "[HIDDEN]",
  "skip_certVerify": false,
  "use_internal_addr": true,
  "adapter": "Trivy",
  "vendor": "Aqua Security",
  "version": "v0.20.1",
  "create_time": "2022-01-14T02:27:45.68801Z",
  "update_time": "2022-01-14T02:27:45.688012Z"
}
2022-01-14T05:23:26Z [DEBUG] [/pkg/scan/job.go:374]: scanRequest:
2022-01-14T05:23:26Z [INFO] [/pkg/scan/job.go:385]: {
  "registry": {
    "url": "https://core:8443",
    "authorization": "[HIDDEN]"
  },
  "artifact": {
    "namespace_id": 2,
    "repository": "test/goharbor/nginx-photon",
    "tag": "sha256-cf18799559139d889acc3c460b20a6b58fa336c7f32e0a172e3e8892619bff3c.sig",
    "digest": "sha256:7db390490e03461a8e1f644406751f70de0473f9a8686a63fd5fa8eb1b5ffee4",
    "mime_type": "application/vnd.oci.image.manifest.v1+json"
  }
}
2022-01-14T05:23:26Z [INFO] [/pkg/scan/job.go:167]: Report mime types: [application/vnd.security.vulnerability.report; version=1.1]
2022-01-14T05:23:26Z [INFO] [/pkg/scan/job.go:222]: Get report for mime type: application/vnd.security.vulnerability.report; version=1.1
2022-01-14T05:23:28Z [DEBUG] [/pkg/scan/job.go:235]: check scan report for mime application/vnd.security.vulnerability.report; version=1.1 at 2022/01/14 05:23:28
2022-01-14T05:23:28Z [INFO] [/pkg/scan/job.go:243]: Report with mime type application/vnd.security.vulnerability.report; version=1.1 is not ready yet, retry after 5 seconds
2022-01-14T05:23:33Z [DEBUG] [/pkg/scan/job.go:235]: check scan report for mime application/vnd.security.vulnerability.report; version=1.1 at 2022/01/14 05:23:33
2022-01-14T05:23:33Z [ERROR] [/pkg/scan/job.go:292]: check scan report with mime type application/vnd.security.vulnerability.report; version=1.1: running trivy wrapper: running trivy: exit status 1: 2022-01-14T05:23:29.100Z	[35mDEBUG[0m	Severities: UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL
2022-01-14T05:23:29.101Z	[35mDEBUG[0m	cache dir:  /home/scanner/.cache/trivy
2022-01-14T05:23:29.101Z	[35mDEBUG[0m	DB Schema: 1, Type: 1, UpdatedAt: 2020-04-17 00:09:46.510533811 +0000 UTC, NextUpdate: 2020-04-17 12:09:46.510533411 +0000 UTC, DownloadedAt: 0001-01-01 00:00:00 +0000 UTC
2022-01-14T05:23:29.101Z	[35mDEBUG[0m	Vulnerability type:  [os library]
2022-01-14T05:23:29.178Z	[35mDEBUG[0m	Image ID: sha256:e5fdf1492d8b9f716510bd816b4762bf6de15ba0481a664f2b6f7118ae6e99f9
2022-01-14T05:23:29.178Z	[35mDEBUG[0m	Diff IDs: [sha256:8b439357eebeb81ada0e6dc78ae39605d8cb8c4773dc024e6c7656a647b040bd]
2022-01-14T05:23:29.178Z	[35mDEBUG[0m	Missing image ID: sha256:e5fdf1492d8b9f716510bd816b4762bf6de15ba0481a664f2b6f7118ae6e99f9
2022-01-14T05:23:29.178Z	[35mDEBUG[0m	Missing diff ID: sha256:8b439357eebeb81ada0e6dc78ae39605d8cb8c4773dc024e6c7656a647b040bd
2022-01-14T05:23:29.186Z	[31mFATAL[0m	scan error:
    github.com/aquasecurity/trivy/pkg/commands/artifact.runWithTimeout
        /home/runner/work/trivy/trivy/pkg/commands/artifact/run.go:71
  - image scan failed:
    github.com/aquasecurity/trivy/pkg/commands/artifact.scan
        /home/runner/work/trivy/trivy/pkg/commands/artifact/run.go:216
  - failed analysis:
    github.com/aquasecurity/trivy/pkg/scanner.Scanner.ScanArtifact
        /home/runner/work/trivy/trivy/pkg/scanner/scan.go:98
  - analyze error:
    github.com/aquasecurity/fanal/artifact/image.Artifact.Inspect
        /home/runner/go/pkg/mod/github.com/aquasecurity/fanal@v0.0.0-20211014152324-29fbacfbfca4/artifact/image/image.go:105
  - failed to analyze layer: sha256:8b439357eebeb81ada0e6dc78ae39605d8cb8c4773dc024e6c7656a647b040bd :
    github.com/aquasecurity/fanal/artifact/image.Artifact.inspect.func1
        /home/runner/go/pkg/mod/github.com/aquasecurity/fanal@v0.0.0-20211014152324-29fbacfbfca4/artifact/image/image.go:155
  - unable to get uncompressed layer sha256:8b439357eebeb81ada0e6dc78ae39605d8cb8c4773dc024e6c7656a647b040bd:
    github.com/aquasecurity/fanal/artifact/image.Artifact.inspectLayer
        /home/runner/go/pkg/mod/github.com/aquasecurity/fanal@v0.0.0-20211014152324-29fbacfbfca4/artifact/image/image.go:195
  - failed to get the layer content (sha256:8b439357eebeb81ada0e6dc78ae39605d8cb8c4773dc024e6c7656a647b040bd):
    github.com/aquasecurity/fanal/artifact/image.Artifact.uncompressedLayer
        /home/runner/go/pkg/mod/github.com/aquasecurity/fanal@v0.0.0-20211014152324-29fbacfbfca4/artifact/image/image.go:262
  - gzip: invalid header
: general response handler: unexpected status code: 500, expected: 200
```

--- a/src/controller/scan/base_controller.go
+++ b/src/controller/scan/base_controller.go
@@ -34,6 +34,7 @@ import (
 	"github.com/goharbor/harbor/src/lib/orm"
 	"github.com/goharbor/harbor/src/lib/q"
 	"github.com/goharbor/harbor/src/lib/retry"
+	"github.com/goharbor/harbor/src/pkg/accessory"
 	allowlist "github.com/goharbor/harbor/src/pkg/allowlist/models"
 	"github.com/goharbor/harbor/src/pkg/permission/types"
 	"github.com/goharbor/harbor/src/pkg/robot/model"
@@ -94,6 +95,8 @@ type basicController struct {
 	manager report.Manager
 	// Artifact controller
 	ar ar.Controller
+	// Accessory manager
+	acc accessory.Manager
 	// Scanner controller
 	sc sc.Controller
 	// Robot account controller
@@ -121,6 +124,8 @@ func NewController() Controller {
 		manager: report.NewManager(),
 		// Refer to the default artifact controller
 		ar: ar.Ctl,
+		// Refer to the default accessory manager
+		acc: accessory.Mgr,
 		// Refer to the default scanner controller
 		sc: sc.DefaultController,
 		// Refer to the default robot account controller
@@ -171,6 +176,14 @@ func (bc *basicController) collectScanningArtifacts(ctx context.Context, r *scan
 	)
 
 	walkFn := func(a *ar.Artifact) error {
+		ok, err := bc.isAccessory(ctx, a)
+		if err != nil {
+			return err
+		}
+		if ok {
+			return nil
+		}
+
 		supported := hasCapability(r, a)
 
 		if !supported && a.IsImageIndex() {
@@ -1047,6 +1060,17 @@ func (bc *basicController) getLatestTagOfArtifact(ctx context.Context, artifactI
 	}
 
 	return tags[0].Name, nil
+}
+
+func (bc *basicController) isAccessory(ctx context.Context, art *ar.Artifact) (bool, error) {
+	ac, err := bc.acc.List(ctx, q.New(q.KeyWords{"ArtifactID": art.Artifact.ID, "digest": art.Artifact.Digest}))
+	if err != nil {
+		return false, err
+	}
+	if len(ac) > 0 {
+		return true, nil
+	}
+	return false, nil
 }
 
 func getArtifactID(extraAttrs map[string]interface{}) int64 {

--- a/src/controller/scan/base_controller_test.go
+++ b/src/controller/scan/base_controller_test.go
@@ -19,6 +19,8 @@ import (
 	"encoding/base64"
 	"encoding/json"
 	"fmt"
+	accessoryModel "github.com/goharbor/harbor/src/pkg/accessory/model"
+	accessorytesting "github.com/goharbor/harbor/src/testing/pkg/accessory"
 	"testing"
 	"time"
 
@@ -58,6 +60,7 @@ type ControllerTestSuite struct {
 	suite.Suite
 
 	artifactCtl         *artifacttesting.Controller
+	accessoryMgr        *accessorytesting.Manager
 	originalArtifactCtl artifact.Controller
 
 	tagCtl *tagtesting.FakeController
@@ -257,6 +260,7 @@ func (suite *ControllerTestSuite) SetupSuite() {
 	params[sca.JobParameterRobot] = robotJSON
 
 	suite.ar = &artifacttesting.Controller{}
+	suite.accessoryMgr = &accessorytesting.Manager{}
 
 	suite.tagCtl = &tagtesting.FakeController{}
 	suite.tagCtl.On("List", mock.Anything, mock.Anything, mock.Anything).Return(nil, nil)
@@ -270,6 +274,7 @@ func (suite *ControllerTestSuite) SetupSuite() {
 		ar:      suite.ar,
 		sc:      sc,
 		rc:      rc,
+		acc:     suite.accessoryMgr,
 		tagCtl:  suite.tagCtl,
 		uuid: func() (string, error) {
 			return "the-uuid-123", nil
@@ -307,6 +312,7 @@ func (suite *ControllerTestSuite) TestScanControllerScan() {
 	}
 
 	{
+		mock.OnAnything(suite.accessoryMgr, "List").Return([]accessoryModel.Accessory{}, nil).Once()
 		// success
 		mock.OnAnything(suite.ar, "Walk").Return(nil).Run(func(args mock.Arguments) {
 			walkFn := args.Get(2).(func(*artifact.Artifact) error)
@@ -328,6 +334,7 @@ func (suite *ControllerTestSuite) TestScanControllerScan() {
 	}
 
 	{
+		mock.OnAnything(suite.accessoryMgr, "List").Return([]accessoryModel.Accessory{}, nil).Once()
 		// delete old report failed
 		mock.OnAnything(suite.ar, "Walk").Return(nil).Run(func(args mock.Arguments) {
 			walkFn := args.Get(2).(func(*artifact.Artifact) error)
@@ -344,6 +351,7 @@ func (suite *ControllerTestSuite) TestScanControllerScan() {
 	}
 
 	{
+		mock.OnAnything(suite.accessoryMgr, "List").Return([]accessoryModel.Accessory{}, nil).Once()
 		// a previous scan process is ongoing
 		mock.OnAnything(suite.ar, "Walk").Return(nil).Run(func(args mock.Arguments) {
 			walkFn := args.Get(2).(func(*artifact.Artifact) error)
@@ -415,6 +423,7 @@ func (suite *ControllerTestSuite) TestScanControllerGetReport() {
 
 // TestScanControllerGetSummary ...
 func (suite *ControllerTestSuite) TestScanControllerGetSummary() {
+	mock.OnAnything(suite.accessoryMgr, "List").Return([]accessoryModel.Accessory{}, nil).Once()
 	mock.OnAnything(suite.ar, "Walk").Return(nil).Run(func(args mock.Arguments) {
 		walkFn := args.Get(2).(func(*artifact.Artifact) error)
 		walkFn(suite.artifact)
@@ -514,6 +523,8 @@ func (suite *ControllerTestSuite) TestScanAll() {
 			"Create", ctx, "SCAN_ALL", int64(0), "SCHEDULE",
 		).Return(executionID, nil).Once()
 
+		mock.OnAnything(suite.accessoryMgr, "List").Return([]accessoryModel.Accessory{}, nil).Once()
+
 		mock.OnAnything(suite.artifactCtl, "List").Return([]*artifact.Artifact{}, nil).Once()
 
 		suite.taskMgr.On("Count", ctx, q.New(q.KeyWords{"execution_id": executionID})).Return(int64(0), nil).Once()
@@ -535,6 +546,8 @@ func (suite *ControllerTestSuite) TestScanAll() {
 		suite.execMgr.On(
 			"Create", ctx, "SCAN_ALL", int64(0), "SCHEDULE",
 		).Return(executionID, nil).Once()
+
+		mock.OnAnything(suite.accessoryMgr, "List").Return([]accessoryModel.Accessory{}, nil).Once()
 
 		mock.OnAnything(suite.artifactCtl, "List").Return([]*artifact.Artifact{suite.artifact}, nil).Once()
 		mock.OnAnything(suite.ar, "Walk").Return(nil).Run(func(args mock.Arguments) {

--- a/src/controller/scan/checker.go
+++ b/src/controller/scan/checker.go
@@ -16,10 +16,11 @@ package scan
 
 import (
 	"context"
-
 	"github.com/goharbor/harbor/src/controller/artifact"
 	"github.com/goharbor/harbor/src/controller/artifact/processor/image"
 	"github.com/goharbor/harbor/src/controller/scanner"
+	"github.com/goharbor/harbor/src/lib/q"
+	"github.com/goharbor/harbor/src/pkg/accessory"
 	models "github.com/goharbor/harbor/src/pkg/scan/dao/scanner"
 )
 
@@ -33,6 +34,7 @@ type Checker interface {
 func NewChecker() Checker {
 	return &checker{
 		artifactCtl:   artifact.Ctl,
+		accMgr:        accessory.Mgr,
 		scannerCtl:    scanner.DefaultController,
 		registrations: map[int64]*models.Registration{},
 	}
@@ -40,6 +42,7 @@ func NewChecker() Checker {
 
 type checker struct {
 	artifactCtl   artifact.Controller
+	accMgr        accessory.Manager
 	scannerCtl    scanner.Controller
 	registrations map[int64]*models.Registration
 }
@@ -69,6 +72,14 @@ func (c *checker) IsScannable(ctx context.Context, art *artifact.Artifact) (bool
 	var scannable bool
 
 	walkFn := func(a *artifact.Artifact) error {
+		ok, err := c.isAccessory(ctx, a)
+		if err != nil {
+			return err
+		}
+		if ok {
+			return nil
+		}
+
 		if hasCapability(r, a) {
 			scannable = true
 			return artifact.ErrBreak
@@ -82,6 +93,17 @@ func (c *checker) IsScannable(ctx context.Context, art *artifact.Artifact) (bool
 	}
 
 	return scannable, nil
+}
+
+func (c *checker) isAccessory(ctx context.Context, art *artifact.Artifact) (bool, error) {
+	ac, err := c.accMgr.List(ctx, q.New(q.KeyWords{"ArtifactID": art.Artifact.ID, "digest": art.Artifact.Digest}))
+	if err != nil {
+		return false, err
+	}
+	if len(ac) > 0 {
+		return true, nil
+	}
+	return false, nil
 }
 
 // hasCapability returns true when scanner has capability for the artifact

--- a/src/controller/scan/checker_test.go
+++ b/src/controller/scan/checker_test.go
@@ -16,6 +16,8 @@ package scan
 
 import (
 	"context"
+	accessoryModel "github.com/goharbor/harbor/src/pkg/accessory/model"
+	accessorytesting "github.com/goharbor/harbor/src/testing/pkg/accessory"
 	"testing"
 
 	"github.com/goharbor/harbor/src/controller/artifact"
@@ -34,10 +36,12 @@ type CheckerTestSuite struct {
 func (suite *CheckerTestSuite) new() *checker {
 	artifactCtl := &artifacttesting.Controller{}
 	scannerCtl := &scannertesting.Controller{}
+	accessoryMgr := &accessorytesting.Manager{}
 
 	return &checker{
 		artifactCtl:   artifactCtl,
 		scannerCtl:    scannerCtl,
+		accMgr:        accessoryMgr,
 		registrations: map[int64]*scanner.Registration{},
 	}
 }
@@ -68,8 +72,10 @@ func (suite *CheckerTestSuite) TestIsScannable() {
 	}, nil)
 
 	{
+
 		art := &artifact.Artifact{}
 
+		mock.OnAnything(c.accMgr, "List").Return([]accessoryModel.Accessory{}, nil).Once()
 		mock.OnAnything(c.artifactCtl, "Walk").Return(nil).Once().Run(func(args mock.Arguments) {
 			walkFn := args.Get(2).(func(*artifact.Artifact) error)
 			walkFn(art)
@@ -85,6 +91,7 @@ func (suite *CheckerTestSuite) TestIsScannable() {
 		art.Type = "IMAGE"
 		art.ManifestMediaType = supportMimeType
 
+		mock.OnAnything(c.accMgr, "List").Return([]accessoryModel.Accessory{}, nil).Once()
 		mock.OnAnything(c.artifactCtl, "Walk").Return(nil).Once().Run(func(args mock.Arguments) {
 			walkFn := args.Get(2).(func(*artifact.Artifact) error)
 			walkFn(art)

--- a/src/server/v2.0/handler/scan.go
+++ b/src/server/v2.0/handler/scan.go
@@ -17,7 +17,6 @@ package handler
 import (
 	"context"
 	"fmt"
-
 	"github.com/go-openapi/runtime/middleware"
 	"github.com/goharbor/harbor/src/common/rbac"
 	"github.com/goharbor/harbor/src/controller/artifact"

